src/hal/stm32f4/spi.rs

```diff
@@ -0,0 +1,292 @@
+             prescaler_shift: u8) -> Result<Spi, Error> {
+    use hal::stm32f4::peripheral_clock::PeripheralClock;
+
+    let (reg, clock) = match peripheral {
+      Peripheral::Spi1 => (&reg::SPI1, PeripheralClock::SPI1Clock),
+      Peripheral::Spi2 => (&reg::SPI2, PeripheralClock::SPI2Clock),
+      Peripheral::Spi3 => (&reg::SPI3, PeripheralClock::SPI3Clock),
+      Peripheral::Spi4 => (&reg::SPI4, PeripheralClock::SPI1Clock),
+      Peripheral::Spi5 => (&reg::SPI5, PeripheralClock::SPI2Clock),
+      Peripheral::Spi6 => (&reg::SPI6, PeripheralClock::SPI3Clock),
+    };
+
+    clock.enable();
+
+    // set direction
+    reg.cr1.set_receive_only(direction == Direction::RxOnly);
+    reg.cr1.set_bidirectional_data_mode(direction == Direction::Rx
+        || direction == Direction::Tx);
+    reg.cr1.set_bidirectional_output_enable(direction == Direction::Tx);
+
+    // set role
+    reg.cr1.set_master(role as usize != 0);
+    reg.cr1.set_internal_slave_select(role as usize != 0);
+    reg.cr1.set_software_slave_management(true);
+    reg.cr2.set_ss_output_enable(false);
+
+    // set data size and format (MSB or LSB)
+    reg.cr1.set_data_frame_format(data_size as usize != 0);
+    reg.cr1.set_frame_format(format as usize != 0);
+
+    // set baud rate
+    if prescaler_shift<1 || prescaler_shift>8 {
+      return Err(Error::BaudRate)
+    }
+    reg.cr1.set_baud_rate(prescaler_shift as u16 - 1);
+
+    // set clock mode
+    reg.cr1.set_clock_phase(ClockPhase::Edge1 as usize != 0);
+    reg.cr1.set_clock_polarity(ClockPolarity::Low as usize != 0);
+
+    reg.i2s_cfgr.set_enable(false);
+    reg.cr1.set_hardware_crc_enable(false);
+    reg.crc.set_polynomial(0); //TODO
+
+    if reg.sr.mode_fault() {
+      Err(Error::Mode)
+    }else {
+      reg.cr1.set_spi_enable(true);
+      Ok(Spi {
+        reg: reg,
+      })
+    }
+  }
+
+  /// Returns the status byte.
+  pub fn get_status(&self) -> u8 {
+    //self.reg.sr.raw() //TODO(kvark): #245 doesn't work
+    let mut r = 0u8;
+    if self.reg.sr.receive_buffer_not_empty() {
+      r |= 1u8<<0;
+    }
+    if self.reg.sr.transmit_buffer_empty() {
+      r |= 1u8<<1;
+    }
+    if self.reg.sr.channel_side() {
+      r |= 1u8<<2;
+    }
+    if self.reg.sr.underrun_flag() {
+      r |= 1u8<<3;
+    }
+    if self.reg.sr.crc_error() {
+      r |= 1u8<<4;
+    }
+    if self.reg.sr.mode_fault() {
+      r |= 1u8<<5;
+    }
+    if self.reg.sr.overrun_flag() {
+      r |= 1u8<<6;
+    }
+    if self.reg.sr.busy_flag() {
+      r |= 1u8<<7;
+    }
+    r
+  }
+}
+
+impl ::hal::spi::Spi for Spi {
+  fn write(&self, value: u8) {
+    wait_for!(self.reg.sr.transmit_buffer_empty());
+    self.reg.dr.set_data(value as u16);
+  }
+
+  fn read(&self) -> u8 {
+    wait_for!(self.reg.sr.receive_buffer_not_empty());
+    self.reg.dr.data() as u8
+  }
+}
+
+mod reg {
+  use volatile_cell::VolatileCell;
+  use core::ops::Drop;
+
+  ioregs!(SPI = {
```

Comment:
Can you please use placement ioregs ( `SPI @ <address>` ) in here and get rid of extern block below?